Rewrite summarize_answers to be more accurate

AdamWill opened the pull request:
The summaries are getting pretty inaccurate in certain cases. Incomplete tests aren't really "missing", but they're treated as such in the summary. Problems with gating.yaml files are counted as failed tests, when they really aren't. The special handling of errors (we kinda count them as both failures and errors) doesn't really make much sense.

This rewrites summarize_answers entirely to give more accurate summaries, with logic that's hopefully clearer, and with only one iteration over the answers. It's fairly similar to the javascript code in Bodhi which does much the same work (we have to duplicate this in Bodhi so it can create a single combined summary for the results of multiple greenwave queries, when it has to run more than one query).

AdamWill commented:
Note: I think there may be a concern that users may be parsing the summary and might be thrown off by changes in it. I know Bodhi *used* to do this, but it no longer does and hasn't for a while (I changed that behaviour myself). I'm not sure if any RH-internal users do it, though.

hluk commented on greenwave/policies.py:
```diff
@@ -512,23 +486,61 @@ def summarize_answers(answers):
+    errored_test_count = 0
+    incomplete_test_count = 0
+    missing_test_count = 0
+    for answer in answers:
```
Instead of using `isinstance()` and having to extend the branches below when adding new Answer subclass, it would be better to wrap the counters in an object and let each answer object update it:

```python
summary = Summary()
for answer in answers:
    answer.summarize(summary)
return summary.as_text()
```

hluk reviewed (commented):
The summaries in `docs` need to be updated.

I hope this won't break any usage - "summary" is meant to be read by users or displayed in UIs but not parsed. (Actually, it breaks our integration tests for gating, but that can be easily fixed.)

Well, the exact format of the summary is not documented anyway. 🤷

(BTW, the CI here is currently failing because of some recent breaking changes in waiverdb. We are working on a fix.)

AdamWill commented:
Updated the summaries in `docs` and implemented the other suggestion.

AdamWill commented:
btw, is greenwave's minimum supported python version high enough that we can switch to `f""` format strings? I've seen enough curly braces for this lifetime...:D

hluk commented:
> btw, is greenwave's minimum supported python version high enough that we can switch to `f""` format strings? I've seen enough curly braces for this lifetime...:D

Yes, definitely! Python 3.9 is what the container image uses.

hluk commented:
There are two failed tests currently:

```
  ____________________ test_make_a_decision_on_queued_result _____________________
  
  requests_session = <requests.sessions.Session object at 0x7f30a5bf2220>
  greenwave_server = 'http://localhost:8080/'
  testdatabuilder = <conftest.TestDataBuilder object at 0x7f30a5c6edc0>
  
      def test_make_a_decision_on_queued_result(requests_session, greenwave_server, testdatabuilder):
          nvr = testdatabuilder.unique_nvr()
          result = testdatabuilder.create_result(item=nvr,
                                                 testcase_name=TASKTRON_RELEASE_CRITICAL_TASKS[0],
                                                 outcome='QUEUED')
          data = {
              'decision_context': 'bodhi_update_push_stable',
              'product_version': 'fedora-26',
              'subject_type': 'koji_build',
              'subject_identifier': nvr,
          }
          r = requests_session.post(greenwave_server + 'api/v1.0/decision', json=data)
          assert r.status_code == 200
          res_data = r.json()
          assert res_data['policies_satisfied'] is False
          assert 'taskotron_release_critical_tasks' in res_data['applicable_policies']
          assert 'taskotron_release_critical_tasks_with_blocklist' in res_data['applicable_policies']
          expected_summary = 'Of 3 required test(s), 3 result(s) missing'
  >       assert res_data['summary'] == expected_summary
  E       AssertionError: assert 'Of 3 require...lt(s) missing' == 'Of 3 require...lt(s) missing'
  E         - Of 3 required test(s), 3 result(s) missing
  E         ?                        ^
  E         + Of 3 required test(s), 1 test(s) incomplete, 2 result(s) missing
  E         ?                        ^^^^^^^^^^^^^^^^^^^^^^^
  
  functional-tests/test_api_v1.py:441: AssertionError
  ____________________ test_make_a_decision_on_running_result ____________________
  
  requests_session = <requests.sessions.Session object at 0x7f30a5bf2220>
  greenwave_server = 'http://localhost:8080/'
  testdatabuilder = <conftest.TestDataBuilder object at 0x7f30a5c6edc0>
  
      def test_make_a_decision_on_running_result(requests_session, greenwave_server, testdatabuilder):
          nvr = testdatabuilder.unique_nvr()
          result = testdatabuilder.create_result(item=nvr,
                                                 testcase_name=TASKTRON_RELEASE_CRITICAL_TASKS[0],
                                                 outcome='RUNNING')
          data = {
              'decision_context': 'bodhi_update_push_stable',
              'product_version': 'fedora-26',
              'subject_type': 'koji_build',
              'subject_identifier': nvr,
          }
          r = requests_session.post(greenwave_server + 'api/v1.0/decision', json=data)
          assert r.status_code == 200
          res_data = r.json()
          assert res_data['policies_satisfied'] is False
          assert 'taskotron_release_critical_tasks' in res_data['applicable_policies']
          assert 'taskotron_release_critical_tasks_with_blocklist' in res_data['applicable_policies']
          expected_summary = 'Of 3 required test(s), 3 result(s) missing'
  >       assert res_data['summary'] == expected_summary
  E       AssertionError: assert 'Of 3 require...lt(s) missing' == 'Of 3 require...lt(s) missing'
  E         - Of 3 required test(s), 3 result(s) missing
  E         ?                        ^
  E         + Of 3 required test(s), 1 test(s) incomplete, 2 result(s) missing
```

AdamWill commented:
Oh, sorry, I thought I'd caught all of those. Will fix.

AdamWill commented:
OK, in both cases there the behaviour is correct and I got the test expectation wrong, so I just fixed the test expectations (and rebased). Should be good now.

mvalik commented on greenwave/policies.py:
```diff
@@ -471,35 +495,26 @@ def to_json(self):
+        if self.non_test_msgs:
+            msgstr = ", ".join("{} {}".format(v, k) for (k, v) in self.non_test_msgs.items())
+        if self.test_msgs:
+            addmsg = "Of {} required test(s), ".format(test_count)
```
"Of {} required test{}, ".format(test_count, "s" * (test_count != 1))

mvalik commented on greenwave/policies.py:
```diff
@@ -471,35 +495,26 @@ def to_json(self):
+            msgstr = ", ".join("{} {}".format(v, k) for (k, v) in self.non_test_msgs.items())
+        if self.test_msgs:
+            addmsg = "Of {} required test(s), ".format(test_count)
+            addmsg += ", ".join("{} {}".format(v, k) for (k, v) in self.test_msgs.items())
```
`addmsg += ", ".join(k.format(num=v, s="s" * {v > 1}) for (k, v) in self.test_msgs.items())`

mvalik commented on greenwave/policies.py:
```diff
@@ -471,35 +495,26 @@ def to_json(self):
+    def to_text(self, test_count):
+        msgstr = ""
+        if self.non_test_msgs:
+            msgstr = ", ".join("{} {}".format(v, k) for (k, v) in self.non_test_msgs.items())
```
`msgstr = ", ".join(k.format(num=v, s="s" * {v > 1}) for (k, v) in self.non_test_msgs.items())`

mvalik commented on greenwave/policies.py:
```diff
@@ -471,35 +495,26 @@ def to_json(self):
+        if self.test_msgs:
+            addmsg = "Of {} required test(s), ".format(test_count)
+            addmsg += ", ".join("{} {}".format(v, k) for (k, v) in self.test_msgs.items())
+            if msgstr:
```
`msgstr = (msgstr + ". ") * bool(msgstr) + addmsg`

hluk commented:
Any update? I'm OK with keeping the `(s)` in the messages for any count (though without it the messages may be a bit clearer).

hluk commented:
> @hluk what do you think of the cuteness required to detect whether to add the s in some cases? It's clever, but I don't like being _overly_ clever when it's not really necessary...

I'm OK with keeping the `(s)` in messages to indicate both single and multiple items.

AdamWill commented:
OK, well, then, there's not really a lot to change, then. The only other thing @mvalik suggests is to rejig slightly how to handle the number of each message (put it in the message template), but...that doesn't really seem like an obvious improvement to me.

I tweaked it to use f-strings instead of .format, and use clearer variable names in the `to_text()` method. I don't really see anything else to do, if we're not going to do the cleverness about adding the "s" or not.

AdamWill commented:
Of course, we could merge this and @mvalik could send their proposed tweaks as a follow-up PR, if they want?

hluk commented:
@mvalik Any thoughts?

BTW, I would like to avoid changing this too often since various tests in our internal infrastructure verify the summary and I'm not sure if anyone else depends on the old format.

mvalik commented:
I think adding the -s is quite easy, but improves the user experience. If you don't like the way `"s" * (num != 1)`, maybe a better solution will be to define a simple function:
```
def s(num: int) -> str:
    return "s" if num != 1 else ""
```
so you can use it as `f"{num} test{s(num)} failed"`

mvalik commented:
Also I mentioned, that I don't like an abstract message "gating.yaml issues". IMO it is much better to mention a particular problem, so the user could know what went wrong and how to deal with it.

hluk commented:
I have marked this as Draft so it does not pop up in our open pull request notifications all the time.

I will add a task in out backlog to see if we can improve the messages further and fix our tests (hopefully in the next sprint).

mvalik commented:
Closing this PR, since changes was merged in the another one.
